bitswap/network/httpnet/httpnet.go

```diff
@@ -0,0 +1,378 @@
+	for _, r := range bsnet.receivers {
+		r.PeerConnected(p.ID)
+	}
+	return nil
+}
+
+func (bsnet *httpnet) DisconnectFrom(ctx context.Context, p peer.ID) error {
+	return nil
+}
+
+func (bsnet *httpnet) Stats() network.Stats {
+	return network.Stats{
+		MessagesRecvd: atomic.LoadUint64(&bsnet.stats.MessagesRecvd),
+		MessagesSent:  atomic.LoadUint64(&bsnet.stats.MessagesSent),
+	}
+}
+
+func (nbsnet *httpnet) TagPeer(p peer.ID, tag string, w int) {
+}
+func (bsnet *httpnet) UntagPeer(p peer.ID, tag string) {
+}
+
+func (bsnet *httpnet) Protect(p peer.ID, tag string) {
+}
+func (bsnet *httpnet) Unprotect(p peer.ID, tag string) bool {
+	return false
+}
+
+func (bsnet *httpnet) NewMessageSender(ctx context.Context, p peer.ID, opts *network.MessageSenderOpts) (network.MessageSender, error) {
+	log.Debugf("NewMessageSender: %s", p)
+	pi := bsnet.host.Peerstore().PeerInfo(p)
+	urls := network.ExtractURLsFromPeer(pi)
+	if len(urls) == 0 {
+		return nil, ErrNoHTTPAddresses
+	}
+
+	return &httpMsgSender{
+		// ctx ??
+		peer:      p,
+		urls:      urls,
+		client:    bsnet.client,
+		receivers: bsnet.receivers,
+		closing:   make(chan struct{}, 1),
+		// opts: todo
+	}, nil
+}
+
+type httpMsgSender struct {
+	client    *http.Client
+	peer      peer.ID
+	urls      []*url.URL
+	receivers []network.Receiver
+	opts      network.MessageSenderOpts
+	closing   chan struct{}
+	closeOnce sync.Once
+}
+
+func (sender *httpMsgSender) SendMsg(ctx context.Context, msg bsmsg.BitSwapMessage) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	bsresp := msg.Clone()
+
+	go func() {
+		for range sender.closing {
+			cancel()
+		}
+	}()
+
+	sendErrors := func(err error) {
+		for _, recv := range sender.receivers {
+			recv.ReceiveError(err)
+		}
+	}
+
+	sendURL, _ := url.Parse(sender.urls[0].String())
+	sendURL.RawQuery = "format=raw&dag-scope=block"
+
+	// TODO: assuming we don't have to manage making concurrent
+	// requests here.
+	for _, entry := range msg.Wantlist() {
+		var method string
+		switch {
+		case entry.Cancel:
+			continue // todo: handle cancelling ongoing
+		case entry.WantType == pb.Message_Wantlist_Block:
+			method = "GET"
+		case entry.WantType == pb.Message_Wantlist_Have:
+			method = "HEAD"
+		default:
+			continue
+		}
+
+		sendURL.Path = "/ipfs/" + entry.Cid.String()
+		headers := make(http.Header)
+		headers.Add("Accept", "application/vnd.ipld.raw")
+
+		req, err := http.NewRequestWithContext(ctx,
+			method,
+			sendURL.String(),
+			nil,
+		)
+		if err != nil {
+			log.Error(err)
+			break
+		}
+		req.Header = headers
+		log.Debugf("cid request to %s %s", method, sendURL)
+		resp, err := sender.client.Do(req)
+		if err != nil { // abort talking to this host
+			log.Error(err)
+			// send error?
+			break
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			if entry.SendDontHave {
+				bsresp.AddDontHave(entry.Cid)
+			}
+			continue
+		}
+
+		// TODO: fixme: limited reader
+		body, err := io.ReadAll(resp.Body)
```

Comment:
I guess for per-block fetch, we would set 2MiB limit to match bitswap's [maxMessageSize](https://github.com/ipfs/boxo/blob/03edbcdd8492c11467dbb0d743524ed883689484/bitswap/client/internal/messagequeue/messagequeue.go#L28)?

fwiw we have prior art limiter  in https://github.com/ipfs/boxo/blob/v0.25.0/routing/http/client/transport.go

Reply:
In principle since we perform the request from a single place and only one type of request we don't need to hijack at the transport and can simply wrap the body with a limited reader.

Reply:
> I guess for per-block fetch, we would set 2MiB limit to match bitswap's [maxMessageSize](https://github.com/ipfs/boxo/blob/03edbcdd8492c11467dbb0d743524ed883689484/bitswap/client/internal/messagequeue/messagequeue.go#L28)?

Yes, we could download blocks any size but if we want those blocks to later be bitswappable, they need to be limited to 2MiB. It is starting to look as if this should be more a blockstore limit rather than appear in multiple places.

Reply:
In fact `maxMessageSize` just ensures no more blocks in a message, but the block could be bigger it seems? (it technically is the size of the bitswap message protobuf, so not only the block itself).? That means the ultimate limit is libp2p `MessageSizeMax`. Anyways, 2MiB sounds reasonable to me as default.